vike/node/plugin/shared/readGitignore.ts

```diff
@@ -0,0 +1,46 @@
+export { getGitignoreLines, readAndSetGitignoreLines }
+
+import { readFile, readdir } from 'node:fs/promises'
+import { join, resolve } from 'node:path'
+import { getGlobalObject, unique } from '../utils.js'
+
+const globalObject = getGlobalObject<{
+  lines: string[]
+}>('readGitignore.ts', { lines: [] })
+
+function getGitignoreLines(excludeNegative = true) {
+  if (excludeNegative) {
+    return globalObject.lines.filter((line) => !line.startsWith('!'))
+  }
+  return globalObject.lines
+}
+
+async function readAndSetGitignoreLines(userRootDir: string) {
+  let currentDir = userRootDir
+  const allLines: string[] = []
+  for (;;) {
+    const res = await readdir(currentDir)
+    const gitignoreFile = res.find((fname) => fname.toLowerCase() === '.gitignore')
+    if (gitignoreFile) {
+      const content = await readFile(join(currentDir, gitignoreFile), 'utf8')
+      const lines = content
+        .replace(/\r\n/g, '\n')
+        .split('\n')
+        .filter((line) => line.length)
+      allLines.push(...lines)
```

Comment:
I see another problem unfortunately: collecting all lines of all .gitignore files isn't correct actually, because where the .gitignore file is located plays a role. Let me elaborate:

```
# /foo/.gitignore
/bar
!/bar/baz
```

means actually:

```
/foo/bar
!/foo/bar/baz
```

Each line should be kind of concatenated with the folder in which the `.gitignore` is located (here `/foo`), but then of course there is a spacial handling for:
* question marks `!`
* leading or no leading `/`

So in other words, globbing all .gitignore in a repo and aggregating all their contents is quite a complex endeavor. Is there a library out there we could use for this?

> Note: and another problem I just noticed is that it doesn't seem that you are removing `# comments` from the .gitignore files? or did I miss it?

Reply:
Sorry I misread the code. I thought you were globbing all `.gitignore` files. But we support only a `.gitignore` in the root folder?

Reply:
Alternatively, instead of not perfectly supporting `.gitignore` and leading to problems and surprises for the user, we could support a very simple `.vikeignore` file instead and say:
* no comments
* no `!`
* only one in the whole repo, at the root of the repo.

What would you think of that @brillout @nitedani ?

Reply:
Yes, if supporting `.gitgnore` (or only a subset) turns out to be too complex, then I guess a new config is the only option. (Feel free to object but I think a config defined in `vite.config.js` instead of `.vikeignore` would be better.)

The neat thing of using `.gitignore` is that it "just works" for the user, doubling down on Vike's zero-config nature. But, yea, if it's too complex / time consuming to implement then let's go for a new config.

So I think the bottom line here is: is supporting `.gitignore` too complex / time consuming?

Reply:
Also note that there are edge cases where using `.gitignore` won't be what the user wants, for example imagine the following setup:

```
# .gitignore
/pages/autogenerated
```

```
// package.json
{
  "scripts": {
    "build": "generate /pages/autogenerated && vite build",
// ...
```

Obviously in this case the user won't be happy that the pages in `/pages/autogenerated` aren't found by vike.

I vote for going away from the idea of parsing `.gitignore` and instead having our own `ignoreDirs` (or similar) vite config

Reply:
I did think of this, but only a handful of users are doing that. My thinking was that, in the long term and depending on what needs arise, we can have both a config `crawl: { skipGitignore: boolean, ignore: string[] }` as well as applying `.gitignore` (or only a subset of it) by default.

For example, a simple subset would be to only apply trivial patterns from `.gitignore`:

```
# .gitignore
dist/ # <-- this one is picked
/some/precise/path/to/data/ # <-- this one is picked as well
/foo/**/boo # <-- Ignore? Or maybe we can pick this as one as well while checking if a `!` line negates it
```

Maybe, with a new Vitest `.spec.ts` suite of unit tests, this can be made reliable?

@nitedani WDYT?

Reply:
Yes, I tried to write an algorithm to parse all of the gitignore files in the monorepo, and while it worked, it would be too heavy to maintain. So I think we should only use a subset of the gitignore in config.root.